File: src/domain/recurring.ts
import type { Category, Transaction } from './entities/types'
import type { RecurringCandidate, RecurringItem } from './entities/recurring'
import { normalizeText } from './normalize'
import { assertHalalas, sumMoney } from './money'
import { daysBetween, daysInMonth, formatIsoDate, parseIsoDate } from './period'

export function shiftMonths(date: string, delta: number): string {
  const p = parseIsoDate(date)
  const total = p.year * 12 + p.month - 1 + delta
  const year = Math.floor(total / 12), month = ((total % 12) + 12) % 12 + 1
  return formatIsoDate({ year, month, day: Math.min(p.day, daysInMonth(year, month)) })
}
export function recurringKey(t: Transaction): string {
  return t.merchantId ? 'id:' + t.merchantId : 'name:' + normalizeText(t.rawMerchantName ?? '')
}
const services = /اشتراك|اتصالات|فواتير|مرافق/
const knownService = /^(netflix|spotify|google one|youtube premium|icloud|stc|mobily|zain)(\b|$)/i

/** Three consecutive cycles, at most one charge per month, suitable service,
 * ±5% amounts and ±7 days. Candidates are never financial entries. */
export function detectRecurring(rows: readonly Transaction[], categories: readonly Category[]): RecurringCandidate[] {
  const names = new Map(categories.map(c => [c.id, c.name]))
  const groups = new Map<string, Transaction[]>()
  for (const t of new Map(rows.map(t => [t.id, t])).values()) {
    if (t.observedDirection !== 'out' || !['purchase', 'unclassified'].includes(t.economicKind)) continue
    const name = t.rawMerchantName?.trim()
    if (!name || !(services.test(names.get(t.categoryId ?? '') ?? t.sourceCategory ?? '') || knownService.test(name))) continue
    const key = recurringKey(t) + '|' + t.currency
    groups.set(key, [...(groups.get(key) ?? []), t])
  }
  const out: RecurringCandidate[] = []
  for (const group of groups.values()) {
    group.sort((a,b) => a.occurredAt.localeCompare(b.occurredAt))
    if (group.length < 3) continue
    const months = group.map(t => t.occurredAt.slice(0,7))
    if (new Set(months).size !== group.length) continue
    const recent = group.slice(-3)
    const p = recent.map(t => parseIsoDate(t.occurredAt))
    const gaps = [1,2].map(i => (p[i].year-p[i-1].year)*12+p[i].month-p[i-1].month)
    const cycle = gaps[0]
    if (![1,3,12].includes(cycle) || gaps[1] !== cycle) continue
    if (recent.slice(1).some((t,i) => Math.abs(daysBetween(shiftMonths(recent[i].occurredAt,cycle),t.occurredAt)) > 7)) continue
    const amounts = recent.map(t => t.amountMinor).sort((a,b) => a-b)
    const median = amounts[1]
    if (median <= 0 || amounts.some(a => !Number.isSafeInteger(a) ||
      BigInt(Math.abs(a-median))*100n > BigInt(median)*5n)) continue
    const last = recent[2]
    out.push({ name: last.rawMerchantName!, merchantKey: recurringKey(last), currency:last.currency,
      expectedMinor:median, nextDueAt:shiftMonths(last.occurredAt,cycle),
      cycleMonths:cycle as 1|3|12, transactionIds:group.map(t=>t.id),
      reason:'٣ دورات متتابعة لخدمة مناسبة، بمبلغ متقارب ±٥٪ وموعد متقارب ±٧ أيام.' })
  }
  return out
}
export function recurringSummary(item: RecurringItem, rows: readonly Transaction[], today: string) {
  const from = shiftMonths(today,-12)
  const matched = rows.filter(t => t.occurredAt > from && t.occurredAt <= today &&
    t.currency === item.currency && t.observedDirection === 'out' &&
    recurringKey(t) === item.merchantKey && ['purchase','unclassified'].includes(t.economicKind))
  const unique = [...new Map(matched.map(t => [t.id,t])).values()]
  const annual = item.expectedMinor * (12 / item.cycleMonths)
  assertHalalas(annual)
  return { paidMinor:item.merchantKey.startsWith('manual:') ? null : sumMoney(unique.map(t=>t.amountMinor)), paidCount:unique.length,
    annualMinor:annual, overdue:item.active && item.nextDueAt < today }
}
export function validateRecurring(item: RecurringItem) {
  if (!item.name.trim() || item.name.length > 120) throw new Error('اكتب اسم الخدمة، بحد أقصى ١٢٠ حرف.')
  if (![1,3,12].includes(item.cycleMonths)) throw new Error('اختار دورة شهرية أو كل ٣ شهور أو سنوية.')
  if (!['subscription','bill'].includes(item.kind)) throw new Error('نوع الالتزام غير صالح.')
  if (!['SAR','EGP','USD','EUR','GBP','AED'].includes(item.currency)) throw new Error('العملة غير مدعومة.')
  assertHalalas(item.expectedMinor)
  if (item.expectedMinor <= 0) throw new Error('قيمة الدورة لازم تكون أكبر من صفر.')
  parseIsoDate(item.nextDueAt)
  assertHalalas(item.expectedMinor*(12/item.cycleMonths))
  if(!item.merchantKey||item.merchantKey.length>240) throw new Error('ربط الخدمة غير صالح.')
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Recurring.kt
package app.masroufy.core

import kotlin.math.abs

/**
 * الاشتراكات والفواتير الدورية — نقل `src/domain/recurring.ts` + الكيانات.
 * الاقتراح: ٣ دورات متتابعة، دفعة واحدة في الشهر بالكتير، خدمة مناسبة، المبلغ ±٥٪ والموعد ±٧ أيام.
 * الاقتراح **مش عملية مالية**.
 */
data class RecurringItem(
    val id: String,
    val name: String,
    val merchantKey: String,
    /** "subscription" / "bill". */
    val kind: String,
    /** 1 / 3 / 12. */
    val cycleMonths: Int,
    val expectedMinor: Halalas,
    val currency: Currency,
    val nextDueAt: IsoDate,
    val active: Boolean,
    val confirmed: Boolean,
)

data class RecurringCandidate(
    val name: String,
    val merchantKey: String,
    val currency: Currency,
    val expectedMinor: Halalas,
    val nextDueAt: IsoDate,
    val cycleMonths: Int,
    val transactionIds: List<String>,
    val reason: String,
)

fun shiftMonths(date: IsoDate, delta: Int): IsoDate {
    val p = parseIsoDate(date)
    val total = p.year * 12 + p.month - 1 + delta
    val year = total.floorDiv(12)
    val month = total.mod(12) + 1
    return formatIsoDate(DateParts(year, month, minOf(p.day, daysInMonth(year, month))))
}

fun recurringKey(t: Transaction): String =
    if (!t.merchantId.isNullOrEmpty()) "id:" + t.merchantId else "name:" + normalizeText(t.rawMerchantName ?: "")

private val SERVICES = Regex("اشتراك|اتصالات|فواتير|مرافق")
private val KNOWN_SERVICE = Regex("^(netflix|spotify|google one|youtube premium|icloud|stc|mobily|zain)(${JsText.B}|$)", RegexOption.IGNORE_CASE)
private val RECURRING_KINDS = setOf(EconomicKind.PURCHASE, EconomicKind.UNCLASSIFIED)

fun detectRecurring(rows: List<Transaction>, categories: List<Category>): List<RecurringCandidate> {
    val names = categories.associate { it.id to it.name }
    val groups = LinkedHashMap<String, MutableList<Transaction>>()
    val unique = LinkedHashMap<String, Transaction>().apply { rows.forEach { put(it.id, it) } }
    for (t in unique.values) {
        if (t.observedDirection != Direction.OUT || t.economicKind !in RECURRING_KINDS) continue
        val name = t.rawMerchantName?.let(JsText::trim)
        if (name.isNullOrEmpty()) continue
        val categoryText = names[t.categoryId ?: ""] ?: t.sourceCategory ?: ""
        if (!(SERVICES.containsMatchIn(categoryText) || KNOWN_SERVICE.containsMatchIn(name))) continue
        groups.getOrPut(recurringKey(t) + "|" + t.currency.name) { mutableListOf() }.add(t)
    }
    val out = mutableListOf<RecurringCandidate>()
    for (group0 in groups.values) {
        val group = group0.sortedBy { it.occurredAt }
        if (group.size < 3) continue
        if (group.map { it.occurredAt.take(7) }.toSet().size != group.size) continue
        val recent = group.takeLast(3)
        val p = recent.map { parseIsoDate(it.occurredAt) }
        val gaps = listOf(1, 2).map { i -> (p[i].year - p[i - 1].year) * 12 + p[i].month - p[i - 1].month }
        val cycle = gaps[0]
        if (cycle !in listOf(1, 3, 12) || gaps[1] != cycle) continue
        if ((1..2).any { i -> abs(daysBetween(shiftMonths(recent[i - 1].occurredAt, cycle), recent[i].occurredAt)) > 7 }) continue
        val amounts = recent.map { it.amountMinor }.sorted()
        val median = amounts[1]
        if (median <= 0 || amounts.any { it > MAX_SAFE_HALALAS || abs(it - median) * 100 > median * 5 }) continue
        val last = recent[2]
        out += RecurringCandidate(
            last.rawMerchantName!!, recurringKey(last), last.currency, median, shiftMonths(last.occurredAt, cycle), cycle, group.map { it.id },
            "٣ دورات متتابعة لخدمة مناسبة، بمبلغ متقارب ±٥٪ وموعد متقارب ±٧ أيام.",
        )
    }
    return out
}

data class RecurringSummary(val paidMinor: Halalas?, val paidCount: Int, val annualMinor: Halalas, val overdue: Boolean)

fun recurringSummary(item: RecurringItem, rows: List<Transaction>, today: IsoDate): RecurringSummary {
    val from = shiftMonths(today, -12)
    val matched = rows.filter {
        it.occurredAt > from && it.occurredAt <= today && it.currency == item.currency && it.observedDirection == Direction.OUT &&
            recurringKey(it) == item.merchantKey && it.economicKind in RECURRING_KINDS
    }
    val unique = LinkedHashMap<String, Transaction>().apply { matched.forEach { put(it.id, it) } }.values.toList()
    val annual = item.expectedMinor * (12 / item.cycleMonths)
    assertHalalas(annual)
    return RecurringSummary(
        if (item.merchantKey.startsWith("manual:")) null else sumMoney(unique.map { it.amountMinor }),
        unique.size, annual, item.active && item.nextDueAt < today,
    )
}

class RecurringError(message: String) : IllegalArgumentException(message)

fun validateRecurring(item: RecurringItem) {
    if (JsText.trim(item.name).isEmpty() || item.name.length > 120) throw RecurringError("اكتب اسم الخدمة، بحد أقصى ١٢٠ حرف.")
    if (item.cycleMonths !in listOf(1, 3, 12)) throw RecurringError("اختار دورة شهرية أو كل ٣ شهور أو سنوية.")
    if (item.kind !in listOf("subscription", "bill")) throw RecurringError("نوع الالتزام غير صالح.")
    assertHalalas(item.expectedMinor)
    if (item.expectedMinor <= 0) throw RecurringError("قيمة الدورة لازم تكون أكبر من صفر.")
    parseIsoDate(item.nextDueAt)
    assertHalalas(item.expectedMinor * (12 / item.cycleMonths))
    if (item.merchantKey.isEmpty() || item.merchantKey.length > 240) throw RecurringError("ربط الخدمة غير صالح.")
}
